Snapshot restore tests (#16647)
hashicorp/consul: .changelog/16647.txt

> integration test to cover the use case of taking and restoring a snapshot on a consul cluster

Reviewer: Do we normally add testing items to the changelog? I'm not even quite sure what happens to text outside of the release note.

Reviewer 2: yea, that was a remnant that was not meant to be there.  I removed it.

```release-note:bug

> raft_logstore: Fixes a bug where restoring a snapshot with a WAL backend causes panic because raf-wal expects a monotonically increasing log index and the restore inserts a gap where no files exist.

Reviewer: ```suggestion
raft_logstore: Fixes a bug where restoring a snapshot when using the experimental WAL storage backend causes a panic.
```
nits really, but see what you think

Reviewer 2: I've updated this.  Thanks for the suggestion.
